### PR TITLE
Fixing abs for Rational{<:Unsigned}

### PR DESCRIPTION
**Pull request by lcontento**
I noticed that calling `abs` for values of type `Rational{<:Unsigned}` results in an `OverflowError` being raised. E.g., in Julia 1.0.1 I get
```
julia> abs(one(Rational{UInt}))
ERROR: OverflowError: cannot negate unsigned number
Stacktrace:
 [1] - at ./rational.jl:240 [inlined]
 [2] abs(::Rational{UInt64}) at ./number.jl:144
 [3] top-level scope at none:0
```
Since `abs(x::Real)` in `number.jl` uses `ifelse`, it tries to compute `-x` which is not possible for `x::Rational{<:Unsigned}`. A possible solution for this problem is specializing `abs` like
```
Base.abs(x::Rational{<:Unsigned}) = x
```
as done in this pull request.

### CODE REVIEW

File: base/rational.jl [hunk 1]
```diff
@@ -226,6 +226,8 @@ signbit(x::Rational) = signbit(x.num)
 copysign(x::Rational, y::Real) = copysign(x.num,y) // x.den
 copysign(x::Rational, y::Rational) = copysign(x.num,y.num) // x.den
 
+abs(x::Rational{T}) where {T<:Unsigned} = x
```
Comment (fredrikekre) [inline base/rational.jl]: An alternative approach (and maybe more general?) would be to define
```
abs(x::Rational) = Rational(abs(x.num), x.den)
```
but I don't know what is best.
I also wonder if the `ifelse` in `numbers.jl` could simply be a ternary, i.e.
```
abs(x::Real) = signbit(x) ? -x : x
```
since, as noted in this PR, not all numbers can be negated.

File: test/rational.jl [hunk 2]
```diff
@@ -96,6 +96,10 @@ using Test
     @test !(1//3 < NaN)
     @test !(1//3 == NaN)
     @test !(1//3 > NaN)
+
+    @test abs(one(Rational{UInt})) === one(Rational{UInt})
+    @test abs(one(Rational{Int})) === one(Rational{Int})
+    @test abs(-one(Rational{Int})) === one(Rational{Int})
```
Comment (fredrikekre) [inline test/rational.jl]: Would be good to add a comment to reference this PR just above the tests. The two last tests seems unrelated to the PR?

**Review by fredrikekre — COMMENTED** [review]
Thanks for a great first contribution! Welcome!

File: base/rational.jl (hunk 1 above)
Comment (lcontento) [reply base/rational.jl]: Indeed @fredrikekre's proposal appears to be more general. Should I add another commit to the same PR making the change or resubmit a cleaner one?

By the way, I was using unsigned rationals since it is a possible tag type for EXIF metadata. I agree that they are a strange choice in a general use case.

File: test/rational.jl (hunk 2 above)
Comment (lcontento) [reply test/rational.jl]: You are perfectly right. Just the first test is related to the bug. However, I did not find other tests regarding the abs method, so I thought that checking that the signed version works too would be useful. I admit I have not much experience about writing testcases, so if you think the tests should be kept to a minimum I will remove the last two.

**Comment by lcontento** [comment]
Happy to help! Even if it's just a very small bug ;-)

File: base/rational.jl (hunk 1 above)
Comment (StefanKarpinski) [reply base/rational.jl]: Just add more commits here. Much easier that way. We can squash it when it’s ready.

File: test/rational.jl (hunk 2 above)
Comment (StefanKarpinski) [reply test/rational.jl]: These seem fine to me. The main thing is to test it and I don’t mind having a slightly unrelated test added as well.

**Comment by StefanKarpinski** [comment]
Not at all! It’s great to have it noticed and fixed.

**Review by fredrikekre — APPROVED** [review]

**Review by StefanKarpinski — APPROVED** [review]
CI Failures look unrelated; I've restarted the ones that failed.
